File: legacy/api/routes/human_tasks.py
"""Human-task inbox = Temporal visibility, not a table.

GET lists open GraphflowHumanTask workflows on our task queue carrying this
instance's id prefix, enriched concurrently via the `task_info` query.
POST submits via a synchronous workflow UPDATE: validator rejections come
back as 422 with the reviewer-facing message; the task keeps waiting."""

from __future__ import annotations

import asyncio

from fastapi import APIRouter, Depends, HTTPException, Request
from temporalio.client import WorkflowUpdateFailedError
from temporalio.service import RPCError, RPCStatusCode

from api import schemas
from api.deps import artifact_meta, get_db, or_404
from engine import db as dbm
from engine import runtime

router = APIRouter(tags=["human-tasks"])


@router.get("/human-tasks")
async def list_human_tasks(request: Request, engagement_id: int | None = None) -> list[dict]:
    client = request.app.state.client
    prefix = f"node-{request.app.state.instance}-"
    query = (
        f"TaskQueue = '{runtime.task_queue()}' AND WorkflowType = 'GraphflowHumanTask' "
        "AND ExecutionStatus = 'Running'"
    )
    visible = [wf async for wf in client.list_workflows(query) if wf.id.startswith(prefix)]

    async def enrich(wf) -> dict | None:
        try:
            info = await client.get_workflow_handle(wf.id).query("task_info")
        except Exception:
            return None  # raced to completion, or transient — skip this sweep
        if not info.get("open"):
            return None
        return {
            "task_id": wf.id,
            "engagement_id": info.get("engagement_id"),
            "workflow_id": info.get("workflow_id"),
            "node_id": info.get("node_id"),
            "output_kind": info.get("output_kind"),
            "display_name": info.get("display_name"),
            "instructions": info.get("instructions"),
            "payload": info.get("payload"),
            "result_required_keys": info.get("result_required_keys", []),
            "requested_by_workflow_run": info.get("requested_by_workflow_run"),
            "input_artifact_ids": info.get("input_artifact_ids", []),
            "start_time": wf.start_time.isoformat() if wf.start_time else None,
        }

    tasks = [t for t in await asyncio.gather(*(enrich(wf) for wf in visible)) if t is not None]
    if engagement_id is not None:
        tasks = [t for t in tasks if t["engagement_id"] == engagement_id]
    return tasks


@router.post("/human-tasks/{task_id}/submit")
async def submit_human_task(
    task_id: str,
    body: schemas.HumanTaskSubmit,
    request: Request,
    conn=Depends(get_db),
) -> dict:
    if not task_id.startswith(f"node-{request.app.state.instance}-"):
        raise HTTPException(status_code=404, detail="task not found")
    handle = request.app.state.client.get_workflow_handle(task_id)
    try:
        ref = await handle.execute_update(
            "submit", {"reviewer": body.reviewer, "result": body.result}
        )
    except WorkflowUpdateFailedError as exc:
        # Validator rejection: the reviewer-facing ApplicationError message.
        cause = exc.cause
        message = getattr(cause, "message", None) or str(cause)
        raise HTTPException(status_code=422, detail=message)
    except RPCError as exc:
        if exc.status == RPCStatusCode.NOT_FOUND:
            # Unknown id, or the workflow already completed.
            raise HTTPException(status_code=404, detail="task not found or already completed")
        raise
    art = or_404(dbm.get_artifact, conn, ref["artifact_id"])
    return {"artifact": artifact_meta(art)}

File: backend/src/api/routes/HumanTasks.ts
import type { FastifyInstance } from 'fastify';
import type { ZodTypeProvider } from 'fastify-type-provider-zod';
import { getArtifact } from '../../infrastructure/db/Db.js';
import { NotFoundError } from '../../shared/errors/Errors.js';
import { humanTaskIdPrefix } from '../../temporal/Ids.js';
import type { ApiDeps, TaskWorkflowExecution } from '../Deps.js';
import { withConn } from '../Deps.js';
import type { HumanTaskOut } from '../Schemas.js';
import { HumanTaskSubmitSchema, HumanTasksQuerySchema, TaskIdParamsSchema } from '../Schemas.js';
import type { ArtifactMetaOut } from '../Serializers.js';
import { artifactMeta } from '../Serializers.js';

// The inbox is Temporal visibility, not a table: list Running GraphflowHumanTask workflows on the
// task queue, filter by this db instance's id prefix, enrich concurrently via the task_info query.

export function registerHumanTaskRoutes(app: FastifyInstance, deps: ApiDeps): void {
  const r = app.withTypeProvider<ZodTypeProvider>();
  const prefix = humanTaskIdPrefix(deps.instance);

  const enrich = async (execution: TaskWorkflowExecution): Promise<HumanTaskOut | null> => {
    try {
      const info = await deps.temporal.queryTaskInfo(execution.workflowId);
      // Visibility is eventually consistent; completed tasks can linger as Running.
      if (!info.open) {
        return null;
      }
      return {
        task_id: execution.workflowId,
        engagement_id: info.engagement_id,
        workflow_id: info.workflow_id,
        node_id: info.node_id,
        output_nodeparamslot: info.output_nodeparamslot,
        display_name: info.display_name,
        instructions: info.instructions,
        payload: info.payload,
        result_required_keys: info.result_required_keys,
        requested_by_workflow_run: info.requested_by_workflow_run,
        input_artifact_ids: info.input_artifact_ids,
        start_time: execution.startTime,
      };
    } catch {
      // Raced to completion / transient query failure: drop from this sweep.
      return null;
    }
  };

  r.get(
    '/human-tasks',
    { schema: { querystring: HumanTasksQuerySchema } },
    async (request): Promise<HumanTaskOut[]> => {
      const executions = await deps.temporal.listTaskWorkflows();
      const mine = executions.filter((e) => e.workflowId.startsWith(prefix));
      const enriched = await Promise.all(mine.map(enrich));
      const open = enriched.filter((t): t is HumanTaskOut => t !== null);
      const engagementId = request.query.engagement_id;
      return engagementId === undefined ? open : open.filter((t) => t.engagement_id === engagementId);
    }
  );

  r.post(
    '/human-tasks/:task_id/submit',
    { schema: { params: TaskIdParamsSchema, body: HumanTaskSubmitSchema } },
    async (request): Promise<{ artifact: ArtifactMetaOut }> => {
      const taskId = request.params.task_id;
      // Foreign prefixes are rejected before touching Temporal (shared namespace/queue).
      if (!taskId.startsWith(prefix)) {
        throw new NotFoundError('task not found');
      }
      // The API attributes every submission to a user principal; the body carries the bare name.
      // (A name may itself contain ':' — 'user:' prefixes verbatim, no parsing.)
      const ref = await deps.temporal.executeSubmit(taskId, {
        reviewer: `user:${request.body.reviewer}`,
        result: request.body.result,
      });
      return withConn(deps, (conn) => ({ artifact: artifactMeta(getArtifact(conn, ref.artifact_id)) }));
    }
  );
}
